R/cache-disk.R

```diff
@@ -0,0 +1,533 @@
+#'   deleted.
+#' @param missing A value to return or a function to execute when
+#'   \code{get(key)} is called but the key is not present in the cache. The
+#'   default is a \code{\link{key_missing}} object. If it is a function to
+#'   execute, the function must take one argument (the key), and you must also
+#'   use \code{exec_missing = TRUE}. If it is a function, it is useful in most
+#'   cases for it to throw an error, although another option is to return a
+#'   value. If a value is returned, that value will in turn be returned by
+#'   \code{get()}. See section Missing keys for more information.
+#' @param exec_missing If \code{FALSE} (the default), then treat \code{missing}
+#'   as a value to return when \code{get()} results in a cache miss. If
+#'   \code{TRUE}, treat \code{missing} as a function to execute when
+#'   \code{get()} results in a cache miss.
+#' @param logfile An optional filename or connection object to where logging
+#'   information will be written. To log to the console, use \code{stdout()}.
+#'
+#' @export
+diskCache <- function(
+  dir = NULL,
+  max_size = 10 * 1024 ^ 2,
+  max_age = Inf,
+  max_n = Inf,
+  evict = c("lru", "fifo"),
+  destroy_on_finalize = NULL,
+  missing = key_missing(),
+  exec_missing = FALSE,
+  logfile = NULL)
+{
+  DiskCache$new(dir, max_size, max_age, max_n, evict, destroy_on_finalize,
+                missing, exec_missing, logfile)
+}
+
+
+DiskCache <- R6Class("DiskCache",
+  public = list(
+    initialize = function(
+      dir = NULL,
+      max_size = 10 * 1024 ^ 2,
+      max_age = Inf,
+      max_n = Inf,
+      evict = c("lru", "fifo"),
+      destroy_on_finalize = NULL,
+      missing = key_missing(),
+      exec_missing = FALSE,
+      logfile = NULL)
+    {
+      if (exec_missing && (!is.function(missing) || length(formals(missing)) == 0)) {
+        stop("When `exec_missing` is true, `missing` must be a function that takes one argument.")
+      }
+      if (is.null(destroy_on_finalize)) {
+        destroy_on_finalize <- is.null(dir)
+      }
+      if (is.null(dir)) {
+        dir <- tempfile("DiskCache-")
+      }
+
+      private$dir                 <- absolutePath(dir)
+      private$max_size            <- max_size
+      private$max_age             <- max_age
+      private$max_n               <- max_n
+      private$evict               <- match.arg(evict)
+      private$destroy_on_finalize <- destroy_on_finalize
+      if (private$evict == "lru" && !check_atime_support(private$dir)) {
+        # Another possibility for handling lack of atime support would be to
+        # create a file on disk that contains atimes. However, this would not
+        # be safe when multiple processes are sharing a cache.
+        warning("DiskCache: can't use eviction policy \"lru\" because filesystem for ",
+          private$dir, " does not support atime, or has low atime resolution. Using \"fifo\" instead."
+        )
```

Comment:
It'd be nice to use `touch` (and treating mtime as atime) instead of just falling back to fifo.